### PR TITLE
Distributed Sampling Dataset

### PR DESCRIPTION
**Pull request by warner-benjamin**
This PR copies the NumPy sampling strategy from allenai/OLMo#231 and later PRs to fix the sampling issue with `torch.randperm` in `DistributedSampler`.

As @ohallstrom's analysis showed, the `DistributedSampler` does not sample uniformly when the dataset size is too large. Where large may be as little as 100M samples but almost certainly at 1B. (Rigorous statistical analysis would need to be done to determine the exact point where PyTorch RNG deviates from uniform randomness). 

The NumPy sampler, specifically the [PCG64DXSM](https://numpy.org/doc/stable/reference/random/upgrading-pcg64.html), appears to fix the issue with the PyTorch RNG while more statistically sound than PCG64.

Initial training tests with the new NumPy-based distributed sampler looks good. A few quick statistical tests seem to indicate the NumPy sampler doesn't deviate from uniform sampling.

### CODE REVIEW

**Comment by ohallstrom** [comment]
Thanks for addressing the shuffling issue!

My comments:

Without a length of the dataset we cannot set time unit in DUR or EP when configuring our schedulers, which is a bit of a pain.

Is there a particular reason for not creating a sampler that inherits from DistributedSampler, and then simply overrides the __iter__ function with the same fct as the overridden function, but with numpy random generator and potentially start index? 

Lastly, I see it is noted as TODO, so maybe you already have thought of it: The start index parameter have to be reset to zero when switching to a new epoch, else resuming from a cp or skipping data due to spikes will give problems once we reach a new epoch

**Comment by warner-benjamin** [comment]
> Is there a particular reason for not creating a sampler that inherits from DistributedSampler, and then simply overrides the **iter** function with the same fct as the overridden function, but with numpy random generator and potentially start index?

The original thought was to use OLMo's code since it was battletested, but we decided to switch to the `DistributedSamplerPCG64DXSM` since it's simpler and some quick tests show it's working. Will wait on the 

There is one downside to `DistributedSamplerPCG64DXSM` over `DistributedSamplingDataset`: it uses an order of magnitude more RAM. _f_ we need the memory savings in the future, we can resurface `DistributedSamplingDataset`.

**Comment by ohallstrom** [comment]
Ok, thanks for the explanation! Did not test the code but LGTM 🚀

**Comment by warner-benjamin** [comment]
Merging since the training tests look good.
